dart/common/Extensible.h

```diff
@@ -0,0 +1,217 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_EXTENSIBLE_H_
+#define DART_COMMON_EXTENSIBLE_H_
+
+#include <memory>
+#include <vector>
+
+namespace dart {
+namespace common {
+
+/// Extensible is a CRTP base class that provides an interface for easily
+/// creating data structures that are meant to be extended. Ordinary copying
+/// does not work with extended data structures, because information from the
+/// derived classes will typically be lost during a copy. The Extensible class
+/// provides an interface for creating a new copy of an extended data structure,
+/// as well as copying information between two extended data structures of the
+/// same type.
+template <class T>
+class Extensible
+{
+public:
+
+  /// Default constructor
+  Extensible() = default;
+
+  /// Virtual destructor
+  virtual ~Extensible() = default;
+
+  /// Do not copy this class directly, use clone() or copy() instead
+  Extensible(const Extensible& doNotCopy) = delete;
+
+  /// Do not copy this class directly, use clone() or copy() instead
+  Extensible& operator=(const Extensible& doNotCopy) = delete;
+
+  /// Implement this function to allow your Extensible type to be copied safely.
+  virtual std::unique_ptr<T> clone() const = 0;
+
+  /// Copy the contents of anotherExtensible into this one.
+  virtual void copy(const T& anotherExtensible) = 0;
+};
+
+/// The ExtensibleMixer class is used to easily create an Extensible (such as
+/// Node::State) which simply takes an existing class (Mixin) and creates an
+/// Extensible that wraps it. This creates all the appropriate copy, move, and
+/// clone members, allowing you to follow the Rule Of Zero. You can also
+/// construct an instance in the exact same way that you would construct a Mixin
+/// instance.
+template <class T, class Mixin>
+class ExtensibleMixer : public T, public Mixin
```

Comment:
Yes, you're exactly right: this implements the ``Extensible`` interface for POD structures.

The second argument is used as you would expect: it passes in the POD that ``ExtensibleMixer`` is supposed to wrap.

The first argument is understandably less obvious. There are many different independent components of the code base that use the ``Extensible`` class. Specifically, ``Node::State``, ``Node::Properties``, ``Addon::State``, and ``Addon::Properties`` all inherit ``Extensible``. However, I think it's important to ensure that it's not possible to accidentally input an ``Addon::Properties`` object into something that expects a ``Node::State``, because that would be silly.

So ``Extensible`` defines this interface of extensibility, but we do **not** explicitly deal in terms of ``Extensible`` objects; instead, we deal in terms of object types that have inherited ``Extensible``. For example, we would not want to have a ``std::vector<Extensible>`` because this is too abstract and we wouldn't know if we're looking at a vector of ``Addon::State``, ``Node::Properties``, or something else. Instead we would want to have a ``std::vector<Addon::State>`` or a ``std::vector<Node::Properties>`` so there's no question of what specific kind of ``Extensible`` it is.

For the ``ExtensibleMixer`` class to correctly implement the specific type of Extensible that you intend to use, you must pass in that Extensible type as the first argument. For example:

```
using CustomAddonState = common::ExtensibleMixer<Addon::State, CustomData>;
```

That being said, you don't have to do it like this, because I provide ``Addon::StateMixer``, ``Addon::PropertiesMixer``, ``Node::StateMixer``, and ``Node::PropertiesMixer`` so you can do this instead:

```
using CustomAddonState = common::Addon::StateMixer<CustomData>
```

That way you don't need to worry about that first template argument.

I should probably put a note of this in the doxygen comments for ``ExtensibleMixer``.